### PR TITLE
feat(matrix) : [FOR-436] creation and edition of matrix

### PR DESCRIPTION
**Pull request by qperie**
## Describe your changes
We add a nex type of question : the matrix.
It includes : 
- display this new type in creation pop-up
- the matrix appears in form-editor page
- the matrix is editable
- the matrix can be deleted
- the matrix can be duplicated
- a form containing matrix questions can be duplicated

## Issue ticket number and link
FOR-436 : https://jira.support-ent.fr/browse/FOR-436

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [ ] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: common/src/main/resources/ts/models/FormElement/Question.ts [hunk 1]
```diff
@@ -89,20 +103,33 @@ export class Questions extends Selection<Question> {
...
+            let data = await questionChoiceService.listChoices(this.all.map(q => q.id));
             let listChoices = Mix.castArrayAs(QuestionChoice, data);
             for (let question of this.all) {
-                if (question.question_type === Types.SINGLEANSWER
-                    || question.question_type === Types.MULTIPLEANSWER
-                    || question.question_type === Types.SINGLEANSWERRADIO) {
+                question.choices.all = listChoices.filter(c => c.question_id === question.id);
+                let nbChoices = question.choices.all.length;
+                if (nbChoices <= 0) {
+                    for (let j = 0; j < 3; j++) {
```
Comment (FlorentMr) [inline common/src/main/resources/ts/models/FormElement/Question.ts:114]: Pas possible d'adapter le 3 en une variable? (sinon une constante?)

File: common/src/main/resources/ts/models/FormElement/Question.ts [hunk 2]
```diff
@@ -89,20 +103,33 @@ export class Questions extends Selection<Question> {
...
+    syncChildren = async () : Promise<void> => {
+        let matrixQuestions: Question[] = this.all.filter(q => q.question_type == Types.MATRIX);
+        if (matrixQuestions.length > 0) {
+            let data = await questionService.listChildren(matrixQuestions);
+            let listChildrenQuestions = Mix.castArrayAs(Question, data);
+            for (let question of this.all) {
+                question.children.all = listChildrenQuestions.filter(q => q.matrix_id === question.id);
+                let nbChildren = question.children.all.length;
+                if (nbChildren <= 0) {
+                    for (let j = 0; j < 3; j++) {
```
Comment (FlorentMr) [inline common/src/main/resources/ts/models/FormElement/Question.ts:131]: Pas possible d'adapter le 3 en une variable? (sinon une constante?)

File: formulaire/src/main/resources/public/ts/controllers/form-editor.ts [hunk 3]
```diff
@@ -178,12 +178,16 @@ export const formEditorController = ng.controller('FormEditorController', ['$sco
...
-                if (vm.newElement.question_type === Types.MULTIPLEANSWER
-                    || vm.newElement.question_type === Types.SINGLEANSWER
-                    || vm.newElement.question_type === Types.SINGLEANSWERRADIO) {
+                if (vm.newElement.isTypeChoicesQuestion()) {
                     for (let i = 0; i < 3; i++) {
                         vm.newElement.choices.all.push(new QuestionChoice());
                     }
+
+                    if (vm.newElement.question_type === Types.MATRIX) {
+                        for (let i = 0; i < 3; i++) {
```
Comment (FlorentMr) [inline formulaire/src/main/resources/public/ts/controllers/form-editor.ts:187]: Pas possible d'adapter le 3 en une variable? (sinon une constante?)

**Review by FlorentMr — APPROVED** [review]

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: formulaire/src/main/java/fr/openent/formulaire/controllers/QuestionController.java [hunk 4]
```diff
@@ -61,6 +61,23 @@ public void listForSection(HttpServerRequest request) {
         questionService.listForSection(sectionId, arrayResponseHandler(request));
     }
 
+    @Get("/questions/children")
+    @ApiDoc("List all the children questions of a list of questions")
+    @ResourceFilter(AccessRight.class)
+    @SecuredAction(value = "", type = ActionType.RESOURCE)
+    public void listChildren(HttpServerRequest request) {
+        JsonArray questionIds = new JsonArray();
+        for (Integer i = 0; i < request.params().size(); i++) {
```
[the remark this answers is not in the document]
Comment (qperie) [reply formulaire/src/main/java/fr/openent/formulaire/controllers/QuestionController.java:70]: Non parce que ligne suivante dans le for je fais un request.getParam(i.toString())

**Review by JinDouang — APPROVED** [review]

**Review by FlorentMr — APPROVED** [review]

**Pull request merged**